### ISSUE TITLE
sample + fromPromise depends on order

### ISSUE THREAD

**Issue by maxhoffmann** [cujojs/most#201]
It looks like the order of streams passed to `sample` matters if one or more of them are promises. In the second case `forEach` is never called.

Example:

```js
  describe('most.js/sample-test', () => {
    it('works', (done) => {
      const stream = most.just(1);
      const promise = Promise.resolve(2);
      const sample = most.just(3);

      stream
        .sample((...args) => args, most.fromPromise(promise), sample)
        .forEach(value => assert.ok(value));
    });

    it('does not work', (done) => {
      const stream = most.just(1);
      const promise = Promise.resolve(2);
      const sample = most.just(3);

      stream
        .sample((...args) => args, sample, most.fromPromise(promise)) // order has changed here
        .forEach(value => assert.ok(value));
    });
  })
```

Is this supposed to work like that? I’ve tried it with multiple promises as well. It only works if all promises come first. As soon as one stream comes before one of the promises the stream never emits a value like above.

**Comment by maxhoffmann** [comment]
I’ve created a [requirebin example](http://requirebin.com/?gist=c33460d21d1d96efe868), so that you can reproduce it.

**Comment by briancavalier** [comment]
That certainly is interesting, @maxhoffmann :)  Whenever there are multiple asynchronous mechanisms in play, things can get hairy.  My intuition is that in one case the sampler is ending before the promise fulfillment (or at least before the event that is triggered as a result of the promise's fulfillment arrives), and in the other case, it isn't.  What's not clear is why such a subtle change makes a difference!

Thanks for the requirebin.  I'll take a deeper look soon.

**Comment by briancavalier** [comment]
I've found the issue.  It *is* in the interaction between two asynchronous systems: Promises, which have their own internal scheduling mechanism, and most's Scheduler.  I won't get into details just yet (although it's pretty interesting, imho), but I'm prototyping a solution that looks, well, promising :) and could end up being an overall improvement to `sample()`.

**Comment by briancavalier** [comment]
Also interesting is that I believe the "bug" here is exactly what @maxhoffmann said: that the two argument orderings behave *differently*.  IOW, if the two orderings both resulted in an empty stream (as it does currently with the promise second), I'm not sure I'd classify that as a bug, since promises force all callers to wait for the promise scheduling machinery.  It'd be reasonable just to say that `Promise.resolve(x)` simply "happens after" `just(y)` in all cases.  Or, iow, we'd just say that the `just` sampler always ends before `Promise.resolve` fires.

But since the two orderings do behave differently (and that's entirely unintuitive!), we need to make them behave the same.

**Comment by maxhoffmann** [comment]
I agree, the order of argument types resulting in different behaviours is easy to get wrong and unintuitive. Sounds like you are working on an interesting solution to this problem. Looking forward to see it!

Thanks for taking the time!

**Comment by briancavalier** [comment]
@maxhoffmann I just pushed the [simplify-from-promise](https://github.com/cujojs/most/tree/simplify-from-promise) branch.  Any chance you can give it a shot in a real use case?  It seems to solve the issue in the simple requirebin example (which was immensely helpful, thank you!), but I want to make sure it addresses your situation.

In that branch, all four of these behave identically:

```js
// Given:
var stream = most.just(1);
var promise = Promise.resolve(2);
var sampledStream = most.just(3);

// These are all empty.  IOW, promise parameter order doesn't matter.
stream.sample(Array, most.fromPromise(promise), sampledStream)
stream.sample(Array, sampledStream, most.fromPromise(promise))
stream.sample(Array, sampledStream, most.of(promise).await())
stream.sample(Array, most.of(promise).await(), sampledStream)
```

Interestingly, we had to make a decision about which "happens" first, `just(123)` or `Promise.resolve(123)`.  We decided that `just` "happens before" `Promise.resolve`, at least from most's perspective.  The implication is that using `just(x)` to sample basically anything other than another `just` will always be empty.

**Comment by maxhoffmann** [comment]
@briancavalier Thanks Brian!

Quick feedback: I’ve run the branch against our tests and the one that was failing because of argument order is failing again. (It works with `master` as I moved the promise to the start) Order doesn’t seem to make a difference anymore. It fails either way.

I’ll try to reproduce a minimal test case.

**Comment by maxhoffmann** [comment]
Maybe I didn’t understand your last comment. The streams being empty means they’ll never emit an event?

This is a minimal test case I’d expect to work:

```js
describe('most/sample-with-promise', () => {
  it('should emit an event with promise first', (done) => {
    const stream = most.just(1);
    const sampledStream = most.just(2);
    const promise = Promise.resolve(3);

    stream
      .sample(Array, most.fromPromise(promise), sampledStream)
      .forEach(actual => {
        assert.ok(actual);
        done();
      })
      .catch(done);
  });

  it('should emit an event with stream first', (done) => {
    const stream = most.just(1);
    const sampledStream = most.just(2);
    const promise = Promise.resolve(3);

    stream
      .sample(Array, sampledStream, most.fromPromise(promise))
      .forEach(actual => {
        assert.ok(actual);
        done();
      })
      .catch(done);
  });
});
```

The first one works with `master`. Using `simplify-from-promise` breaks both tests.

**Comment by maxhoffmann** [comment]
@briancavalier Actually I think I’ve used promises for a wrong reason and if both tests fail it’s fine for me.

My basic problem was that I needed to get a value from an already ended stream inside of `flatMap`. I tried using a promise to do this, but it’s better to get the value via `combine` first and then have it in the closure inside of `flatMap`.

At least I found an inconsistent behaviour this way. ;)

**Comment by briancavalier** [comment]
> if both tests fail it’s fine for me

Ok, cool.  I think this turns out to be the most logical, consistent result.

> it’s better to get the value via combine first and then have it in the closure inside of flatMap

Ah, I see.  It's usually better to accumulate the values you will need later, letting those flow through a stream, rather than later trying to reach back into the past.  It sounds like you got to a good solution :+1:

> At least I found an inconsistent behaviour this way

Yes!  I really appreciate it!

I want to do a little more investigation to understand why the change in the `simplify-from-promise` works.  I'm not convinced it's the *best* solution, even though it does appear to have the desired effect!  So, it may be a few days before I can get a patch release out.

**Comment by maxhoffmann** [comment]
Sure, no rush. We don’t use promises anymore so this is not a crucial issue for us.

Thanks for your help!

**Comment by briancavalier** [comment]
I managed to write down why the two argument orderings behaved differently, and also why the change in the branch causes them to have the same behavior.  This is really down in the weeds, but I figured it might be interesting.  It also never hurts to record it somewhere for posterity :)

https://gist.github.com/briancavalier/a27ddc247dd118ae5508

**Comment by maxhoffmann** [comment]
Thanks for the gist! Was an interesting read and I think this is the way to go. Do you plan to merge this soon? Just wondering when I should close this issue. :)

**Comment by briancavalier** [comment]
Thanks for the nudge :)  Merged, and will do a release later today.

**Comment by briancavalier** [comment]
Ok, released [0.18.3](https://github.com/cujojs/most/releases/tag/0.18.3)

**Comment by maxhoffmann** [comment]
Thanks :)
